File: Graphs/bellman_ford_shortest_path.cpp
/**
 * @Author: hardcodecoder
 * @Date:   02:14:46 Thursday 16 April 2020
 * @Last modified by:   hardcodecoder
 * @Last modified time: 06:13:10 Thursday 16 April 2020
 */


#include<stdio.h>
#include<stdlib.h>

//Since infinity is not logically possible in C
//Define infinity to a large value
#define INF 9999

//Struct to represent an edge
struct Edge {
  int src, dest, wt;
};


//Stores the preceding vertex
//from the current vertexx
int *precedingVertexInShortestPath;

//Stores the distance of all
//vertices from source vertex
int *distanceFromSource;


//Method to recursively print
//the path from src --> dest
int printShortestPath(int src, int dest){
  if (src == dest)
    return src;
  printf("%d --> ", printShortestPath(src, precedingVertexInShortestPath[dest]));
  return dest;
}


//Helper method to handle print queries
void print(int v, int src, int dest){
  printf("\nMinimum distance between %d -> %d = %d\n", src, dest, distanceFromSource[dest]);
  printf("\nPath : ");
  printf("%d", printShortestPath(src, dest));

  //Uncomment below code to print all distances from source
  /*printf("\n\nVertex\tDistance from source\n");
  for (int i = 0; i < v; i++){
    int distance = distanceFromSource[i];
    if (distance < 0)
      printf("%d\t\t%d\n", i, distance);
    else
      printf("%d\t\t+%d\n", i, distance);
  }*/
}


//Implementation of Bellman-Ford shortest path algorithm
void bellmanFordShortestPath(int numVertices, int numEdges, struct Edge *edgeList, int src, int dest){
  int i,j;

  distanceFromSource = (int*) malloc(sizeof(int) * numVertices);
  precedingVertexInShortestPath = (int*) malloc(sizeof(int) * numVertices);

  //Initialize all distance from source to infinity
  for (i = 0; i < numVertices; i++)
    distanceFromSource[i] = INF;

  //Initialize distance of source to 0
  distanceFromSource[src] = 0;

  //Relax all the adges for |V| -1 times
  for (i = 0; i < numVertices; i++){
    //Interate over all edges
    for (j = 0; j < numEdges; j++){
      int u = edgeList[j].src;
      int v = edgeList[j].dest;
      int w = edgeList[j].wt;

      //Check new distance from source is
      //lower than previous stored distance
      if (distanceFromSource[u] != INF && distanceFromSource[u] + w < distanceFromSource[v]){
        //Update distance
        distanceFromSource[v] = distanceFromSource[u] + w;
        //Store the preceding vertex
        precedingVertexInShortestPath[v] = u;
      }
    }
  }

  //Relax the edges one more time
  //if the distance(s) changes we have a negative cycle
  //Terminate the program
  for (i = 0; i < numEdges; i++){
    int u = edgeList[i].src;
    int v = edgeList[i].dest;
    int w = edgeList[i].wt;
    if (distanceFromSource[u] != INF && distanceFromSource[u] + w < distanceFromSource[v]){
      //Change in distance, which means a negative cycle
      //Terminate the program
      printf("Negative cycle detected, terminating...");
      return;
    }
  }

  print(numVertices, src, dest);
}


//Driver method
int main(){
  int v, e, src, dest;
  printf("Enter the number of vertices and edges : ");
  scanf("%d %d", &v, &e);

  struct Edge *edgeList = (struct Edge*)malloc(sizeof(struct Edge)*e);
  printf("\nEnter the edge list in the format:\nSrc vertex\tDest vertex\tWt\n");

  for (int i = 0; i < e; i++){
    scanf("%d%d%d", &edgeList[i].src, &edgeList[i].dest, &edgeList[i].wt);
  }

  printf("\nEnter the source and destination vertex : ");
  scanf("%d%d", &src, &dest);

  bellmanFordShortestPath(v, e, edgeList, src, dest);

  /* Test input
  5 8
  0 1 -1
  0 2 4
  1 2 3
  1 3 2
  1 4 2
  3 2 5
  3 1 1
  4 3 -3
  */
  return 0;
}

File: Graphs/Java/BellmanFordShortestPath.java
/**
 * @Author: hardcodecoder
 * @Date:   02:10:35 Friday 15 May 2020
 * @Last modified by:   hardcodecoder
 * @Last modified time: 03:01:57 Friday 15 May 2020
 */


import java.io.*;
import java.util.Scanner;

public class BellmanFordShortestPath {

  // Defines  distance for unreachable vertex
  private static final float INF = 99999;

  // Stores the final shortest distances from source to every other vertex
  // Index represent the vertex no. from source and the value gives the distance
  private float[] shortestDistanceFromSource;

  // Stores the ancesstor to the current vertex
  // Index gives the current vertex no. & value gives the ancesstor vertex no.
  private int[] precedingVertex;

  // main method
  public static void main(String args[]) {
    new BellmanFordShortestPath().initialize();
  }

  // Helper method to input data and initialize necessary variables
  private void initialize() {
    int numVertex, numEdges, mSrc, mDest;

    System.out.println("Enter the number of vertices and edges in the graph: ");
    Scanner sc = new Scanner(System.in);
    numVertex = sc.nextInt();
    numEdges = sc.nextInt();

    // Initialize all necessary variables
    Edge[] edgeList = new Edge[numEdges];
    precedingVertex = new int[numVertex];
    shortestDistanceFromSource = new float[numVertex];

    // Input the edges
    System.out.println("Enter the edge list in the format:\nSrc vertex\tDest vertex\tWt");
    for (int i = 0; i < numEdges; i++){
      int src = sc.nextInt();
      int dest = sc.nextInt();
      float wt = sc.nextFloat();
      edgeList[i] = new Edge(src, dest, wt);
    }

    // Input the source and destination vertex no.
    System.out.print("Enter the source and destination vertex: ");
    mSrc = sc.nextInt();
    mDest = sc.nextInt();

    findShortestpath(numVertex, numEdges, edgeList, mSrc, mDest);
  }

  // Implementation of Bellman-Ford algorithm
  private void findShortestpath(int numVertex, int numEdges, Edge[] edgeList, int source, int destination) {
    // Initialise distance of all vertex from source as Infinity
    for (int i = 0; i < numVertex; i++)
      shortestDistanceFromSource[i] = INF;

    // Set distance of source from source to zero
    shortestDistanceFromSource[source] = 0.0f;

    // Relax all edges for |numVertex -1| times
    for (int i = 0; i < numVertex; i++) {
      for (int j = 0; j < numEdges; j++) {
        // Get data of current edge
        int src = edgeList[j].getSourceVertex();
        int dest = edgeList[j].getDestinationVertex();
        float wt = edgeList[j].getWeight();

        if (shortestDistanceFromSource[src] != INF && shortestDistanceFromSource[src] + wt < shortestDistanceFromSource[dest]) {
          // Distance to 'dest' vertex via 'src' sourtex is better
          shortestDistanceFromSource[dest] = shortestDistanceFromSource[src] + wt;
          precedingVertex[dest] = src;
        }
      }
    }

    // Relax all the edges for one more time
    // to detect presence of negative cycle
    for (int i = 0; i < numEdges; i++) {
      int src = edgeList[i].getSourceVertex();
      int dest = edgeList[i].getDestinationVertex();
      float wt = edgeList[i].getWeight();

      if (shortestDistanceFromSource[src] != INF && shortestDistanceFromSource[src] + wt < shortestDistanceFromSource[dest]) {
        // Negative cycle detected exit
        // without printing shortest path
        System.out.println("Negative cycle detected. Exiting...");
        return;
      }
    }

    print(numVertex, source, destination);
  }

  // Helper method to print the output
  private void print(int numVertex, int src, int dest){
    // Uncomment to print distances of all vertex from source
    /*System.out.println("Shortest distances:\nVertex\tDistance from source\n");
    for (int i = 0; i < numVertex; i++) {
      float d = shortestDistanceFromSource[i];
      if (d < 0)
        System.out.println(String.format("%d\t\t%.1f", i, d));
      else
        System.out.println(String.format("%d\t\t+%.1f", i, d));
    }*/

    // prints the shortest distance
    System.out.print(String.format("Shortest distance from %d to %d: %.2f\nPath: ", src, dest, shortestDistanceFromSource[dest]));
    System.out.println(printPath(src, dest));
  }

  // Method to print the path from source to the destination vertex recursively
  private int printPath(int src, int dest) {
    if (src == dest) return src;
    System.out.print(String.format("%d -- > ", printPath(src, precedingVertex[dest])));
    return dest;
  }

  // Edge object class
  private static class Edge {
    private int sourceVertex, destinationVertex;
    private float weight;

    Edge(int sourceVertex, int destinationVertex, float weight){
      this.sourceVertex = sourceVertex;
      this.destinationVertex = destinationVertex;
      this.weight = weight;
    }

    int getSourceVertex(){
      return sourceVertex;
    }

    int getDestinationVertex(){
      return destinationVertex;
    }

    float getWeight(){
      return weight;
    }
  }
}
